### PR TITLE
5.x: PHPUnit 10

### PR DESCRIPTION
**Pull request by LordSimal**
I thought I would get a little further without creating a draft here but there is already so much I have to document so I don't forget something:

## data provider methods need to be static

I got most of that coverd in this PR but this creates a problem in the `EncryptedCookieMiddlewareTest::malformedCookies` where we use `$this->_encrypt('secret data', 'aes');` from the `CookieCryptTrait.php`
Suggestions?

## `->withConsecutive()` was removed
see https://github.com/sebastianbergmann/phpunit/issues/4564 and https://github.com/sebastianbergmann/phpunit/issues/4026#issuecomment-825453794 

The `->withConsecutive()` removal alone creates a massive load of changes because currently we have 57 usages of this method inside our tests.

I didn't look into that any further till now.

## Multiple methods have been removed without any replacement documentation

- `TestCase::getMockClass()`
- `TestCase::expectWarning()`
- `TestCase::expectWarningMessage()`
- `TestCase::expectWarningMessageMatches()`

and even more on the `Assert` class: https://github.com/sebastianbergmann/phpunit/issues/4602

## `Testcase::run` now returns `void`

I only found the commit where it changed but no mention of why this was changed or how it should be replaced with.
https://github.com/sebastianbergmann/phpunit/commit/77e60150c2185da760692277aeb2b337686b1f30#diff-6c939b18f9a1acd49fa0f01507a01ab61c419136f2d0b94e13fa4e13998389a0R19

We use that inside our [TestCaseTest::testSkipIf()](https://github.com/cakephp/cakephp/blob/5.x/tests/TestCase/TestSuite/TestCaseTest.php#L110)

## Removed Hook System

See https://github.com/sebastianbergmann/phpunit/issues/4595

Which we use in our [PHPUnitExtension](https://github.com/cakephp/cakephp/blob/5.x/src/TestSuite/Fixture/PHPUnitExtension.php#L26)

### CODE REVIEW

**Comment by ADmad** [comment]
There `--verbose` option is gone, seems we need to use `--display-incomplete` and `--display-skipped` instead.

**Comment by LordSimal** [comment]
I have now reduced the amount of failing tests to a minimum as far as I am currently able to.
Not really sure how we should replace `PHPUnitExtension::executeBeforeFirstTest()` because its gone and `TestCase::run()` because it doesn't return a result anymore.

**Comment by Ggizzle** [comment]
It means that you are trying to access it incorrectly. You should never try to access it directly, but rather should make the variables non-static and private and should access them indirectly via non-static public methods and then calling the methods on an instance of the class.

In other words most of your problem is in code you've not shown us -- code where you try to "access" the data.

File: src/TestSuite/Fixture/PHPUnitSubscriber/BeforeFirstTestMethodCalledSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+class BeforeFirstTestMethodCalledSubscriber implements BeforeFirstTestMethodCalledSubscriberInterface
+{
+    /**
+     * Initializes before any tests are run.
+     *
+     * @param \PHPUnit\Event\Test\BeforeFirstTestMethodCalled $event The event
+     * @return void
+     */
+    public function notify(BeforeFirstTestMethodCalled $event): void
```
Comment (ADmad) [inline src/TestSuite/Fixture/PHPUnitSubscriber/BeforeFirstTestMethodCalledSubscriber.php]: This `BeforeFirstTestMethodCalled` event doesn't seem to be triggered if you run phpunit with a test case file as an argument or use the `--filter` option.

Comment (ADmad) [reply src/TestSuite/Fixture/PHPUnitSubscriber/BeforeFirstTestMethodCalledSubscriber.php]: Adding a subscriber for the `PHPUnit\Event\TestSuite\Started` event instead should work.

Comment (ADmad) [reply src/TestSuite/Fixture/PHPUnitSubscriber/BeforeFirstTestMethodCalledSubscriber.php]: https://github.com/dmaicher/doctrine-test-bundle/blob/master/src/DAMA/DoctrineTestBundle/PHPUnit/PHPUnitExtension.php

**Comment by LordSimal** [comment]
I'd appreciate if someone can help me with (or just fix) the rest of the failing tests 😉

**Comment by ADmad** [comment]
> I'd appreciate if someone can help me with (or just fix) the rest of the failing tests :wink:

I fixed couple of failing tests.

No idea what to do for the 1 remaining failure for `TestCaseTest::testSkipIf()`.

**Comment by ADmad** [comment]
> No idea what to do for the 1 remaining failure for TestCaseTest::testSkipIf().

Figured it out :slightly_smiling_face:

File: tests/TestCase/View/Helper/FormHelperTest.php [hunk 2]
```diff
@@ -334,7 +334,7 @@ public function contextSelectionProvider(): array
             ],
         ];
         $form = new Form();
-        $custom = $this->getMockBuilder('Cake\View\Form\ContextInterface')->getMock();
+        //$custom = $this->getMockBuilder('Cake\View\Form\ContextInterface')->getMock();
```
Comment (ADmad) [inline tests/TestCase/View/Helper/FormHelperTest.php]: How do we address this? Providers being static means can't get the mock builder now.

File: tests/TestCase/Http/Middleware/EncryptedCookieMiddlewareTest.php [hunk 3]
```diff
@@ -102,15 +102,15 @@ public function testDecodeMalformedCookies($cookie): void
      *
      * @return array
      */
-    public function malformedCookies(): array
+    public static function malformedCookies(): array
     {
-        $encrypted = $this->_encrypt('secret data', 'aes');
+        //$encrypted = $this->_encrypt('secret data', 'aes');
```
Comment (ADmad) [inline tests/TestCase/Http/Middleware/EncryptedCookieMiddlewareTest.php]: TODO

**Comment by ADmad** [comment]
So another config deprecation was added in phpunit 10.1 :laughing:.

File: phpunit.xml.dist [hunk 4]
```diff
@@ -1,10 +1,9 @@
...
-    colors="true"
-    bootstrap="tests/bootstrap.php"
-    backupGlobals="true"
-    convertDeprecationsToExceptions="true"
-    >
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/10.0/phpunit.xsd"
+         colors="true"
+         bootstrap="tests/bootstrap.php"
+         backupGlobals="true"
+         cacheDirectory=".phpunit.cache">
```
Comment (ADmad) [inline phpunit.xml.dist:6]: @LordSimal Any particular reason for specifying `cacheDirectory` or was it auto added by phpunit's config migrator?

File: phpunit.xml.dist [hunk 5]
```diff
@@ -19,9 +18,9 @@
             <directory>tests/TestCase/Database/</directory>
             <directory>tests/TestCase/ORM/</directory>
         </testsuite>
-        <testsuite name="http-interop-http-factory">
-            <directory>./vendor/http-interop/http-factory-tests/test</directory>
-        </testsuite>
+<!--        <testsuite name="http-interop-http-factory">-->
+<!--            <directory>./vendor/http-interop/http-factory-tests/test</directory>-->
+<!--        </testsuite>-->
```
Comment (othercorey) [inline phpunit.xml.dist:23]: Should remove if this isn't needed.

File: src/TestSuite/Fixture/PHPUnitSubscriber/PHPUnitStartedSubscriber.php [hunk 6]
```diff
@@ -0,0 +1,48 @@
+<?php
```
Comment (othercorey) [inline src/TestSuite/Fixture/PHPUnitSubscriber/PHPUnitStartedSubscriber.php]: Can we just put these into an extension namespace along with the above extension file?

File: phpunit.xml.dist (hunk 5 above)
Comment (ADmad) [reply phpunit.xml.dist:23]: Keeping this comment for now will help us remember to add back the http-interop tests package when they eventually support phpunit 10.

Comment (othercorey) [reply phpunit.xml.dist:23]: I was curious what replaced this.

File: tests/TestCase/ORM/Query/SelectQueryTest.php [hunk 7]
```diff
@@ -1552,7 +1556,7 @@ public function testHydrateHasManyCustomEntity(): void
      */
     public function testHydrateBelongsToCustomEntity(): void
     {
-        $authorEntity = $this->getMockClass('Cake\ORM\Entity', ['foo']);
+        $authorEntity = get_class($this->createMock('Cake\ORM\Entity'));
```
Comment (markstory) [inline tests/TestCase/ORM/Query/SelectQueryTest.php:1559]: Why does createMock() work here, but on 1522 you used an anonymous class?

**Review by markstory — APPROVED** [review]
Looks good to me. Great job finding a solution to `withConsecutive` being removed.

**Comment by righel** [comment]
Hello,
Any plans on making `cakephp/cakephp:4.x` compatible with PHPUnit10 ?

**Comment by LordSimal** [comment]
PHPUnit 10 requires PHP 8.1 which we do not require for CakePHP 4.
~Therefore we can't backport PHPUnit 10.~
We would have to support both PHPUnit 9 and 10.

**Pull request merged**